### PR TITLE
Vtcca 2731

### PR DESCRIPTION
**Pull request by reddy-yattapu**

### CODE REVIEW

File: app/actions/AuthenticatedAction.scala [hunk 1]
```diff
@@ -16,62 +16,57 @@
...
-        Global.onError(request, e)
+        logger.debug("the request passed through business-rates-authorisation")
+        handleResult(res, block)(request, hc)
     }
   }
 
-  def asAgent(body: AgentRequest[AnyContent] => Future[Result])(implicit messages: Messages) = apply { implicit request =>
+  def asAgent(body: AgentRequest[AnyContent] => Future[Result])(implicit messages: Messages) = this.async { implicit request =>
     if (request.organisationAccount.isAgent) {
+      logger.debug("the request was successfully an agent. ")
```
Comment (DanielConnelly) [inline app/actions/AuthenticatedAction.scala]: please remove this log statement.

File: app/actions/AuthenticatedAction.scala [hunk 2]
```diff
@@ -93,19 +88,76 @@ class AuthenticatedAction @Inject()(
...
-                          (implicit request: Request[AnyContent]) = {
+  def success[A](
+                        accounts: Accounts,
+                        body: BasicAuthenticatedRequest[A] => Future[Result])
+                      (implicit request: Request[A], hc: HeaderCarrier): Future[Result] = {
+    def handleError: PartialFunction[Throwable, Future[Result]] = {
+      case _: NoActiveSession =>
+        provider.redirectToLogin
+      case otherException =>
+        Logger.debug(s"Exception thrown on authorisation with message: ${otherException.getMessage}")
```
Comment (DanielConnelly) [inline app/actions/AuthenticatedAction.scala]: can we log the exception, the full stack trace as well.

File: app/assets/sass/modules/_buttons.scss [hunk 3]
```diff
@@ -6,11 +6,17 @@
   cursor: pointer;
   text-decoration: underline;
   color: $link-colour;
+  box-shadow: none;
```
Comment (DanielConnelly) [inline app/assets/sass/modules/_buttons.scss]: why are we adding more custom css.

File: app/connectors/BusinessRatesAttachmentConnector.scala [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import uk.gov.hmrc.http.{HeaderCarrier, HttpErrorFunctions}
+import models.attachment._
+import uk.gov.hmrc.http._
+import uk.gov.hmrc.play.config.ServicesConfig
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.ExecutionContext.Implicits.global
+
+@Singleton
+class BusinessRatesAttachmentConnector @Inject()(val  http: WSHttp)(implicit val appConfig: ServicesConfig, executionContext: ExecutionContext)
```
Comment (DanielConnelly) [inline app/connectors/BusinessRatesAttachmentConnector.scala]: why is servicesConfig coming in implicitly?

File: app/connectors/BusinessRatesAttachmentConnector.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    http.POST[InitiateAttachmentRequest, PreparedUpload](s"$baseURL/business-rates-attachments/initiate", uploadSettings)
+
+  }
+
+  def submitFile(fileReference: String, submissionId: String)(
+    implicit headerCarrier: HeaderCarrier): Future[Option[Attachment]] = {
+    http.PATCH[JsObject, Attachment](s"$baseURL/business-rates-attachments/attachments/${fileReference}",
+      Json.obj(PropertyLinkEvidence.submissionId.toString -> JsString(submissionId))).map(Some.apply).recover {
+      case ex: Exception =>
+        Logger.warn(s"File Submission failed for File Reference: ${fileReference} Response body: ${ex.printStackTrace()}")
```
Comment (DanielConnelly) [inline app/connectors/BusinessRatesAttachmentConnector.scala]: once again can we print stack trace of exception here?

File: app/connectors/BusinessRatesAttachmentConnector.scala [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+import play.api.http.Status.BAD_REQUEST
+import uk.gov.hmrc.http._
+
+trait AttachmentHttpErrorFunctions extends HttpErrorFunctions {
+
+  override def handleResponse(httpMethod: String, url: String)(response: HttpResponse): HttpResponse =
+    response.status match {
+      case BAD_REQUEST =>
+        Logger.warn(s"Upload failed with status ${response.status}. Response body: ${response.body}")
```
Comment (DanielConnelly) [inline app/connectors/BusinessRatesAttachmentConnector.scala]: is this the only case we handle in challenge as well?

File: app/connectors/BusinessRatesAuthorisation.scala [hunk 7]
```diff
@@ -32,8 +33,10 @@ class BusinessRatesAuthorisation @Inject()(config: ServicesConfig, http: WSHttp)
   val url = config.baseUrl("business-rates-authorisation") + "/business-rates-authorisation"
 
 
+  val logger = Logger(this.getClass.getName)
 
   def authenticate(implicit hc: HeaderCarrier): Future[AuthorisationResult] = {
+    logger.debug("authenticate called")
```
Comment (DanielConnelly) [inline app/connectors/BusinessRatesAuthorisation.scala]: can we remove this logging.

File: app/assets/sass/modules/_buttons.scss (hunk 3 above)
Comment (reddy-yattapu) [reply app/assets/sass/modules/_buttons.scss]: We need UpScan File upload render correctly and we have added same for Challenge

File: app/connectors/BusinessRatesAttachmentConnector.scala (hunk 6 above)
Comment (reddy-yattapu) [reply app/connectors/BusinessRatesAttachmentConnector.scala]: Yes, Only we need to exclusively handle  Specific  BAD_REQUEST to file business rate attachment to Display Right error messages on Scan and Rest of cases handle by System Default Handling

File: app/connectors/BusinessRatesAttachmentConnector.scala (hunk 4 above)
Comment (reddy-yattapu) [reply app/connectors/BusinessRatesAttachmentConnector.scala]: Now servicesConfig injecting

File: app/actions/AuthenticatedAction.scala (hunk 2 above)
Comment (reddy-yattapu) [reply app/actions/AuthenticatedAction.scala]: removed

File: app/actions/AuthenticatedAction.scala (hunk 1 above)
Comment (reddy-yattapu) [reply app/actions/AuthenticatedAction.scala]: removed

File: app/connectors/BusinessRatesAttachmentConnector.scala (hunk 5 above)
Comment (reddy-yattapu) [reply app/connectors/BusinessRatesAttachmentConnector.scala]: Do you mean we need to remove this Log statement?

File: app/actions/AuthenticatedAction.scala [hunk 8]
```diff
@@ -16,62 +16,56 @@
...
 
-  override def messagesApi: MessagesApi = messageApi
-
   protected implicit def hc(implicit request: Request[_]): HeaderCarrier = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
 
-  def apply(body: BasicAuthenticatedRequest[AnyContent] => Future[Result]) = Action.async { implicit request =>
+  override def invokeBlock[A](request: Request[A], block: BasicAuthenticatedRequest[A] => Future[Result]): Future[Result] = {
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+    logger.debug("the request called invoke block")
     businessRatesAuthorisation.authenticate flatMap {
```
Comment (habeeb01) [inline app/actions/AuthenticatedAction.scala]: @DanielConnelly 
i thought we were trying to remove BRA ?

File: app/controllers/propertyLinking/Declaration.scala [hunk 9]
```diff
@@ -45,39 +46,25 @@ class Declaration @Inject()(envelopes: EnvelopeConnector,
...
-      form.bindFromRequest().value match {
-        case Some(true) => fileUploads.getFileMetadata(request.ses.envelopeId) flatMap {
-          case data@FileMetadata(_, Some(info)) => submitLinkingRequest(data) map { _ => Redirect(routes.Declaration.confirmation()) }
-          case data@FileMetadata(NoEvidenceFlag, _) => submitLinkingRequest(data) map { _ => Redirect(routes.Declaration.noEvidence()) }
-        }
-        case _ => BadRequest(declaration(DeclarationVM(formWithNoDeclaration)))
-      }
-    } else {
       form.bindFromRequest().value match {
-        case Some(true) => noEvidenceFlag match {
```
Comment (DanielConnelly) [inline app/controllers/propertyLinking/Declaration.scala]: this should be a fold not a match on the .value.

File: app/controllers/propertyLinking/UploadPropertyEvidence.scala [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+     for{
+        - <- businessRatesAttachmentService.persistSessionData(request.ses, request.ses.uploadEvidenceData.copy( attachments = updatedSessionData))
+      }yield (Ok(uploadEvidence(request.ses.submissionId, List.empty, updatedSessionData.getOrElse(Map()), form)))
+  }
+
+
+  def continue() = withLinkingSession { implicit request =>
+    implicit def hc(implicit request: Request[_]) = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+    request.ses.uploadEvidenceData.attachments match {
+      case Some(fileData) if fileData.size > 0 => {
```
Comment (DanielConnelly) [inline app/controllers/propertyLinking/UploadPropertyEvidence.scala]: we know property link is exactly one element right now.

File: app/controllers/propertyLinking/UploadRatesBill.scala [hunk 11]
```diff
@@ -18,61 +18,85 @@ package controllers.propertyLinking
...
+
+  def show() = withLinkingSession { implicit request =>
+    Ok(
+      uploadRatesBill(request.ses.submissionId, List.empty, request.ses.uploadEvidenceData.attachments.getOrElse(Map.empty)))}
+
+
+  def initiate(): Action[JsValue] = authenticated.async(parse.json) { implicit request  =>
+    implicit def hc(implicit request: Request[_]) = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+
+   val parseResult:JsResult[InitiateAttachmentRequest] = request.body.validate[InitiateAttachmentRequest]
```
Comment (DanielConnelly) [inline app/controllers/propertyLinking/UploadRatesBill.scala]: duplicated code!

File: app/models/attachment/SubmissionTypes.scala [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+sealed abstract class SubmissionTypes( val destination: String,
+                               val submissionId: String)
+
+case object SubmissionTypesValues {
+  case object PropertyLinkEvidence extends SubmissionTypes("PROPERTY_LINK_EVIDENCE_DFE", "propertyLinkSubmissionId")
+}
+
+sealed abstract class FileTypes( val value: String)
+case object FileTypes {
+  case object Evidence extends FileTypes("EVIDENCE")
```
Comment (DanielConnelly) [inline app/models/attachment/SubmissionTypes.scala]: copy paste uneeded code.

File: project/FrontendBuild.scala [hunk 13]
```diff
@@ -40,6 +41,7 @@ private object AppDependencies {
     "com.google.guava" % "guava" % "18.0",
     "joda-time" % "joda-time" % "2.8.2",
     "uk.gov.hmrc" %% "frontend-bootstrap" % "12.4.0",
+    "uk.gov.hmrc" %% "bootstrap-play-25" % bootstrapVersion,
```
Comment (DanielConnelly) [inline project/FrontendBuild.scala]: do not run both frontend-bootstrap and bootstrap-play together.

File: test/controllers/VoaPropertyLinkingSpec.scala [hunk 14]
```diff
@@ -16,19 +16,54 @@
...
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
 trait VoaPropertyLinkingSpec extends FlatSpec with MustMatchers with FutureAwaits with DefaultAwaitTimeout with BeforeAndAfterEach
-  with AppendedClues with MockitoSugar with NoMetricsOneAppPerSuite with WSHTTPMock with ScalaFutures {
+  with AppendedClues with MockitoSugar with NoMetricsOneAppPerSuite with WSHTTPMock with ScalaFutures with FakeObjects{
 
   val token = "Csrf-Token" -> "nocheck"
-
+  implicit def materializer: Materializer = app.injector.instanceOf[Materializer]
```
Comment (DanielConnelly) [inline test/controllers/VoaPropertyLinkingSpec.scala]: this is unit level, should have no need for application or Materializer.

File: test/controllers/agent/RepresentationControllerSpec.scala [hunk 15]
```diff
@@ -35,13 +35,16 @@ class RepresentationControllerSpec extends VoaPropertyLinkingSpec {
 
   lazy val request = FakeRequest().withSession(token)
   implicit val hc = HeaderCarrier()
-  
+  val validGroupAccount = arbitrary[GroupAccount].sample.get
```
Comment (DanielConnelly) [inline test/controllers/agent/RepresentationControllerSpec.scala]: can we move away from generators. create your own object here.

File: test/resources/package.scala [hunk 16]
```diff
@@ -317,7 +318,8 @@ package object resources {
     submissionId <- shortString
     personId <- positiveLong
     declaration <- capacityDeclarationGen
-  } yield LinkingSession(address, uarn, envelopeId, submissionId, personId, declaration)
+    uploadEvidenceData <-  Gen.lzy(UploadEvidenceData.empty)
```
Comment (DanielConnelly) [inline test/resources/package.scala]: Gen.const.

File: app/actions/AuthenticatedAction.scala (hunk 8 above)
Comment (reddy-yattapu) [reply app/actions/AuthenticatedAction.scala]: Might do separate Jira ticket , already we have done lot of changes to part of this ticket

File: app/controllers/propertyLinking/UploadPropertyEvidence.scala (hunk 10 above)
Comment (reddy-yattapu) [reply app/controllers/propertyLinking/UploadPropertyEvidence.scala]: This is not Property Link, This is size of files uploaded

File: app/connectors/BusinessRatesAttachmentConnector.scala (hunk 5 above)
Comment (DanielConnelly) [reply app/connectors/BusinessRatesAttachmentConnector.scala]: print stack trace.

Comment (PeterPerhac) [reply app/connectors/BusinessRatesAttachmentConnector.scala]:
```scala
Logger.warn(s"File Submission failed for File Reference: ${fileReference}", ex)
``` 

`warn` can take second argument the throwable that caused the warning

**Pull request merged**